tests/isa-test.cc

```diff
@@ -0,0 +1,774 @@
+	 * probably have no prelude so we can use the address of the function as the
+	 * address of the load. Obviously we can't allow it to be inlined.
+	 */
+	__noinline int *faulting_load(int *volatile *location)
+	{
+		return *location;
+	}
+
+	/**
+	 * Type of (pointer to) Capability -> Capability function that 'filters' a
+	 * capability. `do_load_test`, `do_store_test` and `do_jalr_test` take
+	 * arguments of this type to allow modifying capabilities operands in
+	 * flexible ways before attempting tests.
+	 */
+	template<class T>
+	using CapabilityFilter = Capability<T> (*)(Capability<T>);
+
+	/**
+	 * A CapabilityFilter that simply returns its input i.e. a no-op / identity
+	 * function. Used as default.
+	 */
+	template<class T>
+	Capability<T> identity_filter(Capability<T> input)
+	{
+		return input;
+	}
+
+	/**
+	 * A CapabilityFilter that returns its input with the tag cleared.
+	 */
+	template<class T>
+	Capability<T> clear_tag_filter(Capability<T> input)
+	{
+		input.invalidate();
+		return input;
+	}
+
+	/**
+	 * A CapabilityFilter that returns a sealed capability. Since we don't have
+	 * access to a sealing capability it can't actually seal its input so just
+	 * returns a sentry capability cast to the correct type. This is fine for
+	 * our purposes as we only want it to trigger SealViolations.
+	 */
+	template<class T>
+	Capability<T> get_sealed_capability_filter(Capability<T> inputCapability)
+	{
+		return reinterpret_cast<T *>(get_interrupts_disabled_sentry());
+	}
+
+	/**
+	 * A CapabilityFilter that returns the input capability with permissions
+	 * restricted to the given PermissionSet.
+	 */
+	template<PermissionSet Permissions, class T>
+	Capability<T> permission_filter(Capability<T> inputCapability)
+	{
+		inputCapability.permissions() &= Permissions;
+		return inputCapability;
+	}
+
+	/**
+	 * A CapabilityFilter that returns its input with length set to one.
+	 * This should be sufficient to trigger a BoundsViolation when we use it.
+	 */
+	template<class T>
+	Capability<T> restrict_bounds_filter(Capability<T> inputCapability)
+	{
+		inputCapability.bounds() = 1;
+		return inputCapability;
+	}
+
+	/**
+	 * A CapabilityFilter that returns its input with the address incremented by
+	 * one so that it will be misaligned.
+	 */
+	template<class T>
+	Capability<T> misalign_filter(Capability<T> inputCapability)
+	{
+		Capability<uint8_t> byteCap = {
+		  reinterpret_cast<uint8_t *>(inputCapability.get())};
+		byteCap += 1;
+		return {reinterpret_cast<T *>(byteCap.get())};
+	}
+
+	/**
+	 * Run a test where we attempt a capability load via a capability to a
+	 * global int* after passing it through `baseFilter`. Asserts that we
+	 * receive CHERI fault `expectedFault`.
+	 */
+	void do_load_test(CauseCode               expectedFault,
+	                  CapabilityFilter<int *> baseFilter,
+	                  size_t                  anExpectedMCause = MCauseCheri)
```

Comment:
Very minor nit, but maybe the argument order should be `baseFilter, expectedFault, anExpectedMCause`, so that it's read as something like "I expect a load with this filter to cause this fault and this code"?